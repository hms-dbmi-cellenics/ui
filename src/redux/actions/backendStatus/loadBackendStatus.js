import handleError from 'utils/http/handleError';
import fetchAPI from 'utils/http/fetchAPI';
import {
  BACKEND_STATUS_LOADING,
  BACKEND_STATUS_LOADED,
  BACKEND_STATUS_ERROR,
} from 'redux/actionTypes/backendStatus';
import endUserMessages from 'utils/endUserMessages';

const loadBackendStatus = (experimentId) => async (dispatch) => {
  dispatch({
    type: BACKEND_STATUS_LOADING,
    payload: {
      experimentId,
    },
  });

  const url = `/v1/experiments/${experimentId}/backendStatus`;
  try {
    const status = await fetchAPI(url);
    dispatch({
      type: BACKEND_STATUS_LOADED,
      payload: {
        experimentId,
        status,
      },
    });

    return status;
  } catch (e) {
    const errorMessage = handleError(e, endUserMessages.ERROR_FETCHING_BACKEND_STATUS);

    dispatch({
      type: BACKEND_STATUS_ERROR,
      payload: {
        experimentId,
        error: errorMessage,
      },
    });
  }
};

export default loadBackendStatus;
