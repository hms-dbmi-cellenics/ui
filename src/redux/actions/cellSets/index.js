import createCellSet from './createCellSet';
import deleteCellSet from './deleteCellSet';
import deleteCellClass from './deleteCellClass';
import runCellSetsClustering from './runCellSetsClustering';
import runCellSetsAnnotation from './runCellSetsAnnotation';
import runCellCycleScoring from './runCellCycleScoring';
import updateCellSetsClustering from './updateCellSetsClustering';

import loadCellSets from './loadCellSets';
import reorderCellSet from './reorderCellSet';
import updateCellSetProperty from './updateCellSetProperty';
import updateCellSetSelected from './updateCellSetSelected';
import setCellSetHiddenStatus from './setCellSetHiddenStatus';
import unhideAllCellSets from './unhideAllCellSets';

export {
  createCellSet,
  deleteCellSet,
  deleteCellClass,
  runCellSetsClustering,
  runCellSetsAnnotation,
  updateCellSetsClustering,
  loadCellSets,
  reorderCellSet,
  updateCellSetSelected,
  updateCellSetProperty,
  setCellSetHiddenStatus,
  unhideAllCellSets,
  runCellCycleScoring,
};
