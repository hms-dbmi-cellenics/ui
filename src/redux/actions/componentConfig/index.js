import loadPlotConfig from 'redux/actions/componentConfig/loadPlotConfig';
import loadComponentConfig from 'redux/actions/componentConfig/loadComponentConfig';
import updatePlotConfig from 'redux/actions/componentConfig/updatePlotConfig';
import savePlotConfig from 'redux/actions/componentConfig/savePlotConfig';
import fetchPlotDataWork from 'redux/actions/componentConfig/fetchPlotDataWork';
import updatePlotData from 'redux/actions/componentConfig/updatePlotData';
import resetPlotConfig from 'redux/actions/componentConfig/resetPlotConfig';

export {
  loadPlotConfig,
  loadComponentConfig,
  updatePlotConfig,
  savePlotConfig,
  fetchPlotDataWork,
  updatePlotData,
  resetPlotConfig,
};
