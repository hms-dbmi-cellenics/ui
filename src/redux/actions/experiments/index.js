import createExperiment from './createExperiment';
import saveExperiment from './saveExperiment';
import updateExperiment from './updateExperiment';
import loadExperiments from './loadExperiments';
import switchExperiment from './switchExperiment';
import reorderSamples from './reorderSamples';

export {
  createExperiment,
  saveExperiment,
  updateExperiment,
  loadExperiments,
  switchExperiment,
  reorderSamples,
};
