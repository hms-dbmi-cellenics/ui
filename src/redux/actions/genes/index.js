import loadGeneExpression from './loadGeneExpression';
import changeGeneSelection from './changeGeneSelection';
import loadPaginatedGeneProperties from './loadPaginatedGeneProperties';
import loadMarkerGenes from './loadMarkerGenes';
import loadDownsampledGeneExpression from './loadDownsampledGeneExpression';

export {
  loadGeneExpression,
  changeGeneSelection,
  loadPaginatedGeneProperties,
  loadMarkerGenes,
  loadDownsampledGeneExpression,
};
