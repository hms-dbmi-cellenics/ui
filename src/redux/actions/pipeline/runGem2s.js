import fetchAPI from 'utils/http/fetchAPI';
import handleError from 'utils/http/handleError';
import endUserMessages from 'utils/endUserMessages';
import {
  EXPERIMENT_SETTINGS_PIPELINE_START,
} from 'redux/actionTypes/experimentSettings';

import {
  BACKEND_STATUS_LOADING,
  BACKEND_STATUS_ERROR,
} from 'redux/actionTypes/backendStatus';

import loadBackendStatus from '../backendStatus/loadBackendStatus';

const runGem2s = (experimentId, paramsHash) => async (dispatch) => {
  dispatch({
    type: BACKEND_STATUS_LOADING,
    payload: {
      experimentId,
    },
  });

  const url = `/v1/experiments/${experimentId}/gem2s`;
  try {
    await fetchAPI(
      url,
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ paramsHash }),
      },
      false,
    );

    dispatch({
      type: EXPERIMENT_SETTINGS_PIPELINE_START,
      payload: {},
    });

    dispatch(loadBackendStatus(experimentId));
  } catch (e) {
    let errorMessage = handleError(e, endUserMessages.ERROR_STARTING_PIPLELINE);

    // temporarily give the user more info if the error is permission denied
    if (errorMessage.includes('does not have access to experiment')) {
      errorMessage += ' Refresh the page to continue with your analysis.';
    }

    dispatch({
      type: BACKEND_STATUS_ERROR,
      payload: {
        experimentId,
        error: errorMessage,
      },
    });
  }
};

export default runGem2s;
