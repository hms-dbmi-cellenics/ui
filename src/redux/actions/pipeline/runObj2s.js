import fetchAPI from 'utils/http/fetchAPI';
import handleError from 'utils/http/handleError';
import endUserMessages from 'utils/endUserMessages';
import loadBackendStatus from 'redux/actions/backendStatus/loadBackendStatus';

const runObj2s = (experimentId) => async (dispatch) => {
  try {
    await fetchAPI(
      `/v2/experiments/${experimentId}/seurat`,
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
      },
    );

    await dispatch(loadBackendStatus(experimentId));

    return true;
  } catch (e) {
    const errorMessage = handleError(e, endUserMessages.ERROR_STARTING_PIPLELINE);

    if (errorMessage !== endUserMessages.ERROR_NO_PERMISSIONS) {
      await dispatch(loadBackendStatus(experimentId));
    }

    return false;
  }
};

export default runObj2s;
