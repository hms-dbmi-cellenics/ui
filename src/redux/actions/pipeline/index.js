import runQC from './runQC';
import runGem2s from './runGem2s';
import runSubsetExperiment from './runSubsetExperiment';
import runObj2s from './runObj2s';
import retry from './retry';

export {
  runQC,
  runGem2s,
  runSubsetExperiment,
  runObj2s,
  retry,
};
