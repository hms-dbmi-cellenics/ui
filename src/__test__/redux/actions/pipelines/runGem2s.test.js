import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';
import initialExperimentsState, { experimentTemplate } from 'redux/reducers/experiments/initialState';
import loadBackendStatus from 'redux/actions/backendStatus/loadBackendStatus';
import initialBackendState from 'redux/reducers/backendStatus';

import {
  EXPERIMENT_SETTINGS_PIPELINE_START,
} from 'redux/actionTypes/experimentSettings';

import {
  BACKEND_STATUS_LOADING,
  BACKEND_STATUS_ERROR,
} from 'redux/actionTypes/backendStatus';

import { runGem2s } from 'redux/actions/pipeline';

const mockStore = configureStore([thunk]);
enableFetchMocks();

jest.mock('redux/actions/backendStatus/loadBackendStatus',
  () => jest.fn().mockImplementation(() => async () => { }));

const experimentId = 'experiment-id';
const projectId = 'project-id';
const oldParamsHash = 'old-gem2s-hash';

const initialState = {
  experiments: {
    ...initialExperimentsState,
    [experimentId]: {
      ...experimentTemplate,
      name: 'Mock experiment',
      id: experimentId,
      projectUuid: projectId,
      sampleIds: ['sample-1', 'sample-2'],
    },
  },
  backendStatus: {
    [experimentId]: {
      ...initialBackendState,
      status: {
        gem2s: {
          paramsHash: oldParamsHash,
        },
      },
    },
  },
};

describe('runGem2s action', () => {
  beforeEach(() => {
    const response = new Response(JSON.stringify({}));

    fetchMock.resetMocks();
    fetchMock.doMock();
    fetchMock.mockResolvedValueOnce(response);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('Dispatches events properly', async () => {
    const store = mockStore(initialState);
    await store.dispatch(runGem2s(experimentId));

    const actions = store.getActions();

    expect(actions[0].type).toEqual(BACKEND_STATUS_LOADING);
    expect(actions[1].type).toEqual(EXPERIMENT_SETTINGS_PIPELINE_START);
    expect(loadBackendStatus).toHaveBeenCalled();

    expect(actions).toMatchSnapshot();
  });

  it('Dispatches status error if loading fails', async () => {
    fetchMock.resetMocks();
    fetchMock.mockResponse(JSON.stringify({ message: 'some weird error that happened' }), { status: 400 });

    const store = mockStore(initialState);
    await store.dispatch(runGem2s(experimentId));

    const actions = store.getActions();

    expect(actions[0].type).toEqual(BACKEND_STATUS_LOADING);
    expect(loadBackendStatus).not.toHaveBeenCalled();
    expect(actions[1].type).toEqual(BACKEND_STATUS_ERROR);

    expect(actions).toMatchSnapshot();
  });

  it('Dispatches properly without project data', async () => {
    const store = mockStore(initialState);
    await store.dispatch(runGem2s(experimentId));

    expect(fetchMock).toHaveBeenCalled();
  });
});
