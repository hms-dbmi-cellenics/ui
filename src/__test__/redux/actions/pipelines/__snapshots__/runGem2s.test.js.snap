// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runGem2s action Dispatches events properly 1`] = `
Array [
  Object {
    "payload": Object {
      "experimentId": "experiment-id",
    },
    "type": "backendStatus/backendStatusLoading",
  },
  Object {
    "payload": Object {},
    "type": "experimentSettings/startPipeline",
  },
]
`;

exports[`runGem2s action Dispatches status error if loading fails 1`] = `
Array [
  Object {
    "payload": Object {
      "experimentId": "experiment-id",
    },
    "type": "backendStatus/backendStatusLoading",
  },
  Object {
    "payload": Object {
      "error": "We couldn't start working on your data.",
      "experimentId": "experiment-id",
    },
    "type": "backendStatus/backendStatusError",
  },
]
`;
